File: CHTL_JS/CHTLJSLexer/lexer.py
from CHTL_JS.CHTLJSLexer.token import CHTLJSToken
from CHTL_JS.CHTLJSLexer.token_type import CHTLJSTokenType

class CHTLJSLexer:
    def __init__(self, source: str):
        self.source = source
        self.tokens = []
        self.start = 0
        self.current = 0
        self.line = 1
        self.keywords = {"listen": CHTLJSTokenType.IDENTIFIER} # Can be expanded

    def scan_tokens(self) -> list[CHTLJSToken]:
        while not self._is_at_end():
            self.start = self.current
            self._scan_token()
        self.tokens.append(CHTLJSToken(CHTLJSTokenType.EOF, "", self.line, self.start))
        return self.tokens

    def _scan_token(self):
        char = self._advance()
        if char in ' \r\t\n': return

        if char == '{':
            if self._match('{'): self._add_token(CHTLJSTokenType.LBRACE_LBRACE)
            else: self._add_token(CHTLJSTokenType.LBRACE)
        elif char == '}':
            if self._match('}'): self._add_token(CHTLJSTokenType.RBRACE_RBRACE)
            else: self._add_token(CHTLJSTokenType.RBRACE)
        elif char == '-':
            if self._match('>'): self._add_token(CHTLJSTokenType.ARROW)
        elif char == ':': self._add_token(CHTLJSTokenType.COLON)
        elif char == ';': self._add_token(CHTLJSTokenType.SEMICOLON)
        else:
            # This simplified lexer assumes anything else is an identifier
            # or a placeholder from the Unified Scanner.
            while not self._is_at_end() and self._peek() not in ' \r\t\n{}:;':
                self._advance()
            text = self.source[self.start:self.current]
            # Check for keywords, otherwise it's an identifier/placeholder
            ttype = self.keywords.get(text, CHTLJSTokenType.IDENTIFIER)
            self._add_token(ttype)

    def _is_at_end(self) -> bool: return self.current >= len(self.source)
    def _advance(self) -> str:
        char = self.source[self.current]; self.current += 1; return char
    def _match(self, expected: str) -> bool:
        if self._is_at_end() or self.source[self.current] != expected: return False
        self.current += 1; return True
    def _peek(self) -> str: return '\0' if self._is_at_end() else self.source[self.current]
    def _add_token(self, type: CHTLJSTokenType):
        text = self.source[self.start:self.current]
        self.tokens.append(CHTLJSToken(type, text, self.line, self.start))

File: CHTL_JS/CHTLJSLexer/Lexer.cpp
#include "Lexer.h"

const std::map<std::string, CHTLJSTokenType> CHTLJSLexer::keywords = {
    {"fileloader", CHTLJSTokenType::FileLoader},
    {"listen", CHTLJSTokenType::Listen},
    {"delegate", CHTLJSTokenType::Delegate},
    {"animate", CHTLJSTokenType::Animate},
    {"vir", CHTLJSTokenType::Vir},
    {"router", CHTLJSTokenType::Router},
    {"util", CHTLJSTokenType::Util},
    {"then", CHTLJSTokenType::Then},
    {"iNeverAway", CHTLJSTokenType::INeverAway},
    {"printMylove", CHTLJSTokenType::PrintMyLove}
};

CHTLJSLexer::CHTLJSLexer(const std::string& source) : source(source) {}

std::vector<CHTLJSToken> CHTLJSLexer::tokenize() {
    std::vector<CHTLJSToken> tokens;
    CHTLJSToken token;
    do {
        token = nextToken();
        tokens.push_back(token);
    } while (token.type != CHTLJSTokenType::EndOfFile);
    return tokens;
}

CHTLJSToken CHTLJSLexer::nextToken() {
    skipWhitespace();
    start = current;

    if (isAtEnd()) return makeToken(CHTLJSTokenType::EndOfFile);

    char c = advance();

    switch (c) {
        case '{':
            if (peek() == '{') { advance(); return makeToken(CHTLJSTokenType::DoubleCurlyOpen); }
            return makeToken(CHTLJSTokenType::OpenBrace);
        case '}':
            if (peek() == '}') { advance(); return makeToken(CHTLJSTokenType::DoubleCurlyClose); }
            return makeToken(CHTLJSTokenType::CloseBrace);
        case '(': return makeToken(CHTLJSTokenType::OpenParen);
        case ')': return makeToken(CHTLJSTokenType::CloseParen);
        case '[': return makeToken(CHTLJSTokenType::OpenBracket);
        case ']': return makeToken(CHTLJSTokenType::CloseBracket);
        case ':': return makeToken(CHTLJSTokenType::Colon);
        case ';': return makeToken(CHTLJSTokenType::Semicolon);
        case ',': return makeToken(CHTLJSTokenType::Comma);
        // case '.': return makeToken(CHTLJSTokenType::Dot); // Dot is part of identifiers now
        case '=':
            if (peek() == '>') { advance(); return makeToken(CHTLJSTokenType::FatArrow); }
            return makeToken(CHTLJSTokenType::Equals);
        case '-':
            if (peek() == '>') { advance(); return makeToken(CHTLJSTokenType::Arrow); }
            break; // Fall through to handle as unexpected
        case '"':
        case '\'':
            return stringLiteral(c);
        default:
            if (isAlpha(c)) return identifier();
            if (isDigit(c)) return number();
            // Check for identifiers that don't start with alpha, like ./path
            if (isIdentifierChar(c)) return identifier();
            break;
    }

    return errorToken("Unexpected character.");
}

void CHTLJSLexer::skipWhitespace() {
    while (true) {
        char c = peek();
        switch (c) {
            case ' ':
            case '\r':
            case '\t':
                advance();
                break;
            case '\n':
                line++;
                column = 1;
                advance();
                break;
            default:
                return;
        }
    }
}

CHTLJSToken CHTLJSLexer::identifier() {
    while (isIdentifierChar(peek())) {
        advance();
    }
    std::string text = source.substr(start, current - start);
    auto it = keywords.find(text);
    if (it != keywords.end()) {
        return makeToken(it->second);
    }
    return makeToken(CHTLJSTokenType::Identifier);
}

CHTLJSToken CHTLJSLexer::number() {
    while (isDigit(peek())) {
        advance();
    }
    // Handle floating point
    if (peek() == '.' && isDigit(source[current + 1])) {
        advance(); // consume '.'
        while (isDigit(peek())) {
            advance();
        }
    }
    return makeToken(CHTLJSTokenType::NumberLiteral);
}

CHTLJSToken CHTLJSLexer::stringLiteral(char opening_quote) {
    while (peek() != opening_quote && !isAtEnd()) {
        if (peek() == '\n') line++;
        advance();
    }
    if (isAtEnd()) return errorToken("Unterminated string.");
    advance(); // Closing quote
    return makeToken(CHTLJSTokenType::StringLiteral);
}


CHTLJSToken CHTLJSLexer::makeToken(CHTLJSTokenType type) const {
    return makeToken(type, source.substr(start, current - start));
}

CHTLJSToken CHTLJSLexer::makeToken(CHTLJSTokenType type, const std::string& value) const {
    return CHTLJSToken{type, value, line, (int)(column - (current - start))};
}

CHTLJSToken CHTLJSLexer::errorToken(const std::string& message) const {
    return CHTLJSToken{CHTLJSTokenType::Unexpected, message, line, column};
}

char CHTLJSLexer::peek() {
    if (isAtEnd()) return '\0';
    return source[current];
}

char CHTLJSLexer::advance() {
    if (!isAtEnd()) {
        current++;
        column++;
    }
    return source[current - 1];
}

bool CHTLJSLexer::isAtEnd() const {
    return current >= source.length();
}

bool CHTLJSLexer::isAlpha(char c) const {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}

bool CHTLJSLexer::isDigit(char c) const {
    return c >= '0' && c <= '9';
}

bool CHTLJSLexer::isIdentifierChar(char c) const {
    switch (c) {
        case '{':
        case '}':
        case '(':
        case ')':
        case '[':
        case ']':
        case ':':
        case ';':
        case ',':
        case '=':
        case '\0':
            return false;
        default:
            return !isspace(c);
    }
}
